### PR TITLE
refactor!: remove legacy JS bundle

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

We no longer serve our JavaScript bundle to browsers that do not support modules. Mainly IE11. There are a few exception with inline scripts used for accessibility improvements.

> **Warning**
> Our **Google Analytics** and **Ophan** scripts will no longer function in browsers that do not support ES Modules.

## Why?

We no longer officially support IE11. Let’s improve the experience for no JavaScript instead of outdated JavaScript.

See also https://github.com/guardian/configs/pull/207 for context on current browser usage.

cc. @guardian/client-side-infra

### CODE REVIEW

File: dotcom-rendering/scripts/webpack/webpack.config.browser.js [hunk 1]
```diff
@@ -1,31 +1,27 @@
...
 
 const DEV = process.env.NODE_ENV === 'development';
 
 /**
- * @param {boolean} isLegacyJS
  * @returns {string}
  */
-const generateName = (isLegacyJS) => {
-	const legacyString = isLegacyJS ? '.legacy' : '';
+const generateName = () => {
```
Comment (jamesgorrie) [inline dotcom-rendering/scripts/webpack/webpack.config.browser.js:9]: Does it need to be a `function`?

```ts
const chunkhashString = DEV ? '' : '.[chunkhash]';
const outputFilename = `[name]${chunkhashString}.js`;
```

File: dotcom-rendering/src/web/server/frontToHtml.tsx [hunk 2]
```diff
@@ -102,19 +83,13 @@ export const frontToHtml = ({ front, NAV }: Props): string => {
 	 * unlikely.
 	 */
 	const lowPriorityScriptTags = generateScriptTags([
-		...getScriptArrayFromFile('atomIframe.js'),
-		...getScriptArrayFromFile('embedIframe.js'),
-		...getScriptArrayFromFile('newsletterEmbedIframe.js'),
-		...getScriptArrayFromFile('relativeTime.js'),
+		...getScriptFromFile('atomIframe.js'),
```
Comment (jamesgorrie) [inline dotcom-rendering/src/web/server/frontToHtml.tsx:86]: These can't / shouldn't be `...spread`?

File: dotcom-rendering/docs/architecture/029-only-esmodules.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+## Context
+
+As of June 2022, the Guardian no longer supports IE11.
+
+[cut the mustard]: ./asdf.md
+
+## Decision
+
+Chrome, Edge, Firefox, Opera, Safari all support ES Modules, so we can stop
+serving JavaScript for browsers that don’t [cut the mustard][].
```
Comment (ioannakok) [inline dotcom-rendering/docs/architecture/029-only-esmodules.md:12]: Same as above

File: dotcom-rendering/src/web/server/articleTemplate.ts [hunk 4]
```diff
@@ -196,13 +196,7 @@ https://workforus.theguardian.com/careers/product-engineering/
 
                 <script type="module">
                     window.guardian.mustardCut = true;
-                    window.guardian.gaPath = "${gaPath.modern}";
-                </script>
-
-                <script nomodule>
-                    // Browser fails mustard check
-                    window.guardian.mustardCut = false;
```
Comment (sndrs) [inline dotcom-rendering/src/web/server/articleTemplate.ts:204]: is this ok to remove?

File: dotcom-rendering/src/web/browser/dynamicImport/init.ts [hunk 5]
```diff
@@ -10,36 +10,21 @@ const initialiseDynamicImport = () => {
 		window.guardianPolyfilledImport = new Function(
 			'url',
 			`return import(url)`,
-		) as (url: string) => Promise<any>;
+		) as (url: string) => Promise<unknown>;
```
Comment (ioannakok) [inline dotcom-rendering/src/web/browser/dynamicImport/init.ts:13]: For my understanding, why are we doing this? Is it for safety?

Comment (mxdvl) [reply dotcom-rendering/src/web/browser/dynamicImport/init.ts:13]: By typing as `unknown`, you have to do actual type checking to make the compiler happy.

Probably best to keep out for now, it's not directly related.

File: dotcom-rendering/docs/architecture/029-only-esmodules.md (hunk 3 above)
Comment (oliverlloyd) [reply dotcom-rendering/docs/architecture/029-only-esmodules.md:12]: I think it's going to be helpful to flag this change to future devs as best as possible but just to say that these [decisions docs](https://adr.github.io/) are meant as historical records and - I think - they are not meant to be edited.

I've actually violated that principle myself before but after reading up on this maybe I was wrong to do this. As I understand it, the correct process is to create a new document to record the change in thinking

File: dotcom-rendering/src/lib/assets.ts [hunk 6]
```diff
@@ -34,33 +31,15 @@ const decideAssetOrigin = (stage: string | undefined): string => {
...
-	if (legacyFilename) {
-		scripts.push({
-			src: `${ASSET_ORIGIN}assets/${legacyFilename}`,
-			legacy: true,
-		});
-	}
+	if (!filename) throw new Error(`Could not find filename: ${file}`);
 
-	return scripts;
+	return `${ASSET_ORIGIN}assets/${filename}`;
```
Comment (oliverlloyd) [inline dotcom-rendering/src/lib/assets.ts:44]: 😻 so much clearer!

File: dotcom-rendering/src/web/server/articleTemplate.ts (hunk 4 above)
Comment (oliverlloyd) [reply dotcom-rendering/src/web/server/articleTemplate.ts:204]: 👍 We have `window.guardian.mustardCut` set to true above so I think it's fair to say we should also keep the code that sets it to false.

I think the potential use case here is another team depending on this property. I'm not aware of this being the case but we can't be sure and I don't think we lose anything but having this property available

File: dotcom-rendering/src/web/server/frontTemplate.ts [hunk 7]
```diff
@@ -154,13 +154,7 @@ https://workforus.theguardian.com/careers/product-engineering/
 
                 <script type="module">
                     window.guardian.mustardCut = true;
-                    window.guardian.gaPath = "${gaPath.modern}";
-                </script>
-
-                <script nomodule>
-                    // Browser fails mustard check
-                    window.guardian.mustardCut = false;
```
Comment (oliverlloyd) [inline dotcom-rendering/src/web/server/frontTemplate.ts:162]: As per @sndrs 's comment above, we should keep this negative value to maintain the contract

File: dotcom-rendering/window.guardian.d.ts [hunk 8]
```diff
@@ -16,7 +16,6 @@ declare global {
 	 */
 	interface Window {
 		guardian: {
-			mustardCut: boolean;
```
Comment (oliverlloyd) [inline dotcom-rendering/window.guardian.d.ts:19]: We're still setting this value so I think we need the type as well

**Review by oliverlloyd — APPROVED** [review]
I paired on this so it'd be good to get some other eyes here but I feel happy with the changes. Actually, more than happy, excited. Elated even 🎉

File: dotcom-rendering/src/web/browser/dynamicImport/init.ts (hunk 5 above)
Comment (jamesgorrie) [reply dotcom-rendering/src/web/browser/dynamicImport/init.ts:13]: An extra note here - you can read and write anything to `any`, as it's defined as being able to be anything. It is the most loose type you can get. 

`unknown` is the tightest type you can get, as you can't run most read / write operations on it. e.g:

```typescript
const wat: any = undefined
console.log(wat.things) // valid, but errors on runtime
console.log(wat + 0) // valid, but errors on runtime

const wot: unknown = undefined
console.log(wot.things) // invalid, can't compile
console.log(wot + 0) // invalid, can't compile
```

**Review by jamesgorrie — APPROVED** [review]
Loving it - in [sentry](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/browser/sentryLoader/sentry.ts), should we exclude any errors that come from these browsers as we expect them?

**Comment by sndrs** [comment]
> Loving it - in [sentry](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/browser/sentryLoader/sentry.ts), should we exclude any errors that come from these browsers as we expect them?

Would you expect errors? Should you not serve JS to browsers you know will error?

File: dotcom-rendering/docs/architecture/029-only-esmodules.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Only serve JavaScript ES Modules
+
+## Context
+
+As of June 2022, the Guardian no longer supports IE11.
+
+[cut the mustard]: ./asdf.md
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/docs/architecture/029-only-esmodules.md:7]: I don’t think we want to update the document.

File: dotcom-rendering/docs/architecture/029-only-esmodules.md (hunk 3 above)
Comment (ioannakok) [reply dotcom-rendering/docs/architecture/029-only-esmodules.md:12]: Thanks both for your explanation. Didn't realise we do not update these docs.

File: dotcom-rendering/src/web/browser/dynamicImport/init.ts (hunk 5 above)
Comment (ioannakok) [reply dotcom-rendering/src/web/browser/dynamicImport/init.ts]: Makes sense. Thanks both for your explanations!

File: dotcom-rendering/docs/architecture/029-only-esmodules.md (hunk 3 above)
Comment (mxdvl) [reply dotcom-rendering/docs/architecture/029-only-esmodules.md:12]: I’ve reworked the document and split it in its own commit.

There’s a [live document currently circulating in the department](https://docs.google.com/document/d/1Fhu_g4Q9-a18jWftnGt1n6Scn8x88X-7jX2Ygr8LJrk/edit#), too.

File: dotcom-rendering/src/web/server/frontTemplate.ts (hunk 7 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/server/frontTemplate.ts:162]: It’s back in!

File: dotcom-rendering/window.guardian.d.ts (hunk 8 above)
Comment (mxdvl) [reply dotcom-rendering/window.guardian.d.ts]: Back in!

**Comment by oliverlloyd** [comment]
Are there any more steps needed to get this one over the line? Can I help?

File: dotcom-rendering/scripts/webpack/webpack.config.browser.js [hunk 10]
```diff
@@ -69,25 +60,15 @@ module.exports = ({ isLegacyJS, sessionId }) => ({
...
-													esmodules: true,
-												},
-											},
-									  ],
+								[
+									'@babel/preset-env',
+									{
+										bugfixes: true,
+										targets: {
+											esmodules: true,
```
Comment (sndrs) [inline dotcom-rendering/scripts/webpack/webpack.config.browser.js:68]: how you feel about using https://www.npmjs.com/package/@guardian/browserslist-config here?

**Pull request closed without merging**